src/common/Throttle.h

```diff
@@ -332,23 +332,90 @@ class OrderedThrottle {
   uint32_t waiters = 0;
 };
 
+struct Bucket {
+  CephContext *cct;
+  const std::string name;
 
-class TokenBucketThrottle {
+  uint64_t max = 0;
+  uint64_t rate = 0;
+  uint64_t remain = 0;
 
-  struct Bucket {
-    CephContext *cct;
-    const std::string name;
-    std::atomic<uint64_t> remain = { 0 }, max = { 0 };
+  Bucket(CephContext *cct, const std::string &name, uint64_t max)
+    : cct(cct), name(name), max(max) {}
 
-    Bucket(CephContext *cct, const std::string& n, uint64_t m)
-      : cct(cct), name(n),
-	remain(m), max(m)
-    {
-    }
+  virtual ~Bucket() {};
+
+  /**
+   * Get tokens from Bucket
+   * @param c the value that we want to take
+   * @return actually got
+   */
+  virtual uint64_t get(uint64_t c = 0);
+
+  /**
+   * Put tokens into Bucket
+   * @param c the value that we want to put
+   * @return actually put
+   */
+  virtual uint64_t put(uint64_t c = 0);
 
-    uint64_t get(uint64_t c);
-    uint64_t put(uint64_t c);
-    void set_max(uint64_t m);
+  /**
+   * Set the Bucket size
+   * @param m bucket size
+   * @return true if size is not zero
+   */
+  virtual bool set_max(uint64_t m);
+
+  /**
+   * Set the rate that tokens leave the Bucket
+   * @param r the rate
+   * @return true if rate is not zero
+   */
+  virtual bool set_rate(uint64_t r);
+};
+
+class TokenBucketThrottle {
+
+  struct TokenBucket : Bucket {
```

Comment:
Yes, you got what I said. But should we have a subclass of TokenBucket ? I think the TokenBucketThrottle can take all the work of TokenBucket here. I mean, TokenBucketThrottle can use Bucket and decide how to use it. No need to split the management work into TokenBucketThrottle  and TokenBucket.